### ISSUE TITLE
Clippy v0.0.158 fails to build on nightly (eba374fb2 2017-09-11)

### ISSUE THREAD

**Issue by antifuchs** [rust-lang-nursery/rust-clippy#2044]
I'm running on rust nightly 2017-09-11, and am attempting to install v0.0.158. Current master also fails on the same version (rustc version from `rustup run nightly rustc --version`: `rustc 1.22.0-nightly (eba374fb2 2017-09-11)`):

<details>
<summary>rustup run nightly cargo install clippy --vers 0.0.158 --force</summary>

```
    Updating registry `https://github.com/rust-lang/crates.io-index`
  Installing clippy v0.0.158
   Compiling regex-syntax v0.4.1
   Compiling unicode-normalization v0.1.5
   Compiling lazy_static v0.2.8
   Compiling quine-mc_cluskey v0.2.4
   Compiling quote v0.3.15
   Compiling either v1.1.0
   Compiling unicode-xid v0.0.4
   Compiling num-traits v0.1.40
   Compiling matches v0.1.6
   Compiling itoa v0.3.3
   Compiling dtoa v0.4.2
   Compiling getopts v0.2.15
   Compiling semver-parser v0.7.0
   Compiling pulldown-cmark v0.0.15
   Compiling serde v1.0.14
   Compiling bitflags v0.9.1
   Compiling itertools v0.6.2
   Compiling synom v0.11.3
   Compiling semver v0.6.0
   Compiling syn v0.11.11
   Compiling serde_json v1.0.3
   Compiling toml v0.4.5
   Compiling serde_derive_internals v0.16.0
   Compiling serde_derive v1.0.14
   Compiling clippy_lints v0.0.158
   Compiling cargo_metadata v0.2.3
error[E0308]: mismatched types
   --> /Users/asf/.cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.158/src/utils/mod.rs:944:20
    |
944 |         Def::Local(id) |
    |                    ^^ expected struct `rustc::hir::def_id::DefId`, found struct `syntax::ast::NodeId`
    |
    = note: expected type `rustc::hir::def_id::DefId`
               found type `syntax::ast::NodeId`

error[E0308]: mismatched types
   --> /Users/asf/.cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.158/src/utils/mod.rs:945:20
    |
945 |         Def::Upvar(id, ..) |
    |                    ^^ expected struct `rustc::hir::def_id::DefId`, found struct `syntax::ast::NodeId`
    |
    = note: expected type `rustc::hir::def_id::DefId`
               found type `syntax::ast::NodeId`

error[E0308]: mismatched types
   --> /Users/asf/.cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.158/src/utils/mod.rs:985:34
    |
985 |             path.def.def_id() == defid,
    |                                  ^^^^^ expected struct `rustc::hir::def_id::DefId`, found struct `syntax::ast::NodeId`
    |
    = note: expected type `rustc::hir::def_id::DefId`
               found type `syntax::ast::NodeId`

error[E0308]: mismatched types
   --> /Users/asf/.cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.158/src/utils/inspector.rs:363:65
    |
363 |             if let Some(crate_id) = cx.tcx.extern_mod_stmt_cnum(item.hir_id) {
    |                                                                 ^^^^^^^^^^^ expected struct `rustc::hir::def_id::DefId`, found struct `rustc::hir::HirId`
    |
    = note: expected type `rustc::hir::def_id::DefId`
               found type `rustc::hir::HirId`

error[E0308]: mismatched types
   --> /Users/asf/.cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.158/src/functions.rs:171:14
    |
171 |         Some(def_id)
    |              ^^^^^^ expected struct `rustc::hir::def_id::DefId`, found struct `syntax::ast::NodeId`
    |
    = note: expected type `rustc::hir::def_id::DefId`
               found type `syntax::ast::NodeId`

error[E0308]: mismatched types
  --> /Users/asf/.cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.158/src/let_if_seq.rs:71:35
   |
71 |                 !used_in_expr(cx, def_id, cond),
   |                                   ^^^^^^ expected struct `rustc::hir::def_id::DefId`, found struct `syntax::ast::NodeId`
   |
   = note: expected type `rustc::hir::def_id::DefId`
              found type `syntax::ast::NodeId`

error[E0308]: mismatched types
  --> /Users/asf/.cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.158/src/let_if_seq.rs:73:52
   |
73 |                 let Some(value) = check_assign(cx, def_id, &*then),
   |                                                    ^^^^^^ expected struct `rustc::hir::def_id::DefId`, found struct `syntax::ast::NodeId`
   |
   = note: expected type `rustc::hir::def_id::DefId`
              found type `syntax::ast::NodeId`

error[E0308]: mismatched types
  --> /Users/asf/.cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.158/src/let_if_seq.rs:74:35
   |
74 |                 !used_in_expr(cx, def_id, value),
   |                                   ^^^^^^ expected struct `rustc::hir::def_id::DefId`, found struct `syntax::ast::NodeId`
   |
   = note: expected type `rustc::hir::def_id::DefId`
              found type `syntax::ast::NodeId`

error[E0308]: mismatched types
  --> /Users/asf/.cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.158/src/let_if_seq.rs:80:65
   |
80 |                         if let Some(default) = check_assign(cx, def_id, else_) {
   |                                                                 ^^^^^^ expected struct `rustc::hir::def_id::DefId`, found struct `syntax::ast::NodeId`
   |
   = note: expected type `rustc::hir::def_id::DefId`
              found type `syntax::ast::NodeId`

error[E0308]: mismatched types
   --> /Users/asf/.cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.158/src/loops.rs:805:67
    |
805 |             let manual_copies = get_indexed_assignments(cx, body, def_id);
    |                                                                   ^^^^^^ expected struct `rustc::hir::def_id::DefId`, found struct `syntax::ast::NodeId`
    |
    = note: expected type `rustc::hir::def_id::DefId`
               found type `syntax::ast::NodeId`

error[E0308]: mismatched types
   --> /Users/asf/.cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.158/src/loops.rs:858:22
    |
858 |                 var: def_id,
    |                      ^^^^^^ expected struct `rustc::hir::def_id::DefId`, found struct `syntax::ast::NodeId`
    |
    = note: expected type `rustc::hir::def_id::DefId`
               found type `syntax::ast::NodeId`

error[E0308]: mismatched types
    --> /Users/asf/.cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.158/src/loops.rs:1678:35
     |
1678 |                 .as_local_node_id(def_id)
     |                                   ^^^^^^ expected struct `rustc::hir::def_id::DefId`, found struct `syntax::ast::NodeId`
     |
     = note: expected type `rustc::hir::def_id::DefId`
                found type `syntax::ast::NodeId`

error[E0308]: mismatched types
   --> /Users/asf/.cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.158/src/misc.rs:580:35
    |
580 |                 .as_local_node_id(def_id)
    |                                   ^^^^^^ expected struct `rustc::hir::def_id::DefId`, found struct `syntax::ast::NodeId`
    |
    = note: expected type `rustc::hir::def_id::DefId`
               found type `syntax::ast::NodeId`

error[E0308]: mismatched types
   --> /Users/asf/.cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.158/src/needless_pass_by_value.rs:133:38
    |
133 |                 !moved_vars.contains(&defid),
    |                                      ^^^^^^ expected struct `rustc::hir::def_id::DefId`, found struct `syntax::ast::NodeId`
    |
    = note: expected type `&rustc::hir::def_id::DefId`
               found type `&syntax::ast::NodeId`

error[E0308]: mismatched types
   --> /Users/asf/.cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.158/src/needless_pass_by_value.rs:142:59
    |
142 |                     let deref_span = spans_need_deref.get(&defid);
    |                                                           ^^^^^^ expected struct `rustc::hir::def_id::DefId`, found struct `syntax::ast::NodeId`
    |
    = note: expected type `&rustc::hir::def_id::DefId`
               found type `&syntax::ast::NodeId`

error: aborting due to 15 previous errors

error: Could not compile `clippy_lints`.
warning: build failed, waiting for other jobs to finish...
error: failed to compile `clippy v0.0.158`, intermediate artifacts can be found at `/var/folders/08/j4g_jn953lngpvgmyg8dygk00000gn/T/cargo-install.TGPaDcE67zOw`

Caused by:
  build failed
```
</details>

<details>
<summary>rustup run nightly cargo build (in a checkout of master HEAD)</summary>

```
   Compiling clippy_lints v0.0.158 (file:///Users/asf/Hacks/rust-clippy/clippy_lints)
error[E0308]: mismatched types
   --> clippy_lints/src/utils/mod.rs:953:20
    |
953 |         Def::Local(id) |
    |                    ^^ expected struct `rustc::hir::def_id::DefId`, found struct `syntax::ast::NodeId`
    |
    = note: expected type `rustc::hir::def_id::DefId`
               found type `syntax::ast::NodeId`

error[E0308]: mismatched types
   --> clippy_lints/src/utils/mod.rs:954:20
    |
954 |         Def::Upvar(id, ..) |
    |                    ^^ expected struct `rustc::hir::def_id::DefId`, found struct `syntax::ast::NodeId`
    |
    = note: expected type `rustc::hir::def_id::DefId`
               found type `syntax::ast::NodeId`

error[E0308]: mismatched types
   --> clippy_lints/src/utils/mod.rs:994:34
    |
994 |             path.def.def_id() == defid,
    |                                  ^^^^^ expected struct `rustc::hir::def_id::DefId`, found struct `syntax::ast::NodeId`
    |
    = note: expected type `rustc::hir::def_id::DefId`
               found type `syntax::ast::NodeId`

error[E0308]: mismatched types
   --> clippy_lints/src/utils/inspector.rs:363:65
    |
363 |             if let Some(crate_id) = cx.tcx.extern_mod_stmt_cnum(item.hir_id) {
    |                                                                 ^^^^^^^^^^^ expected struct `rustc::hir::def_id::DefId`, found struct `rustc::hir::HirId`
    |
    = note: expected type `rustc::hir::def_id::DefId`
               found type `rustc::hir::HirId`

error[E0308]: mismatched types
   --> clippy_lints/src/functions.rs:171:14
    |
171 |         Some(def_id)
    |              ^^^^^^ expected struct `rustc::hir::def_id::DefId`, found struct `syntax::ast::NodeId`
    |
    = note: expected type `rustc::hir::def_id::DefId`
               found type `syntax::ast::NodeId`

error[E0308]: mismatched types
  --> clippy_lints/src/let_if_seq.rs:71:35
   |
71 |                 !used_in_expr(cx, def_id, cond),
   |                                   ^^^^^^ expected struct `rustc::hir::def_id::DefId`, found struct `syntax::ast::NodeId`
   |
   = note: expected type `rustc::hir::def_id::DefId`
              found type `syntax::ast::NodeId`

error[E0308]: mismatched types
  --> clippy_lints/src/let_if_seq.rs:73:52
   |
73 |                 let Some(value) = check_assign(cx, def_id, &*then),
   |                                                    ^^^^^^ expected struct `rustc::hir::def_id::DefId`, found struct `syntax::ast::NodeId`
   |
   = note: expected type `rustc::hir::def_id::DefId`
              found type `syntax::ast::NodeId`

error[E0308]: mismatched types
  --> clippy_lints/src/let_if_seq.rs:74:35
   |
74 |                 !used_in_expr(cx, def_id, value),
   |                                   ^^^^^^ expected struct `rustc::hir::def_id::DefId`, found struct `syntax::ast::NodeId`
   |
   = note: expected type `rustc::hir::def_id::DefId`
              found type `syntax::ast::NodeId`

error[E0308]: mismatched types
  --> clippy_lints/src/let_if_seq.rs:80:65
   |
80 |                         if let Some(default) = check_assign(cx, def_id, else_) {
   |                                                                 ^^^^^^ expected struct `rustc::hir::def_id::DefId`, found struct `syntax::ast::NodeId`
   |
   = note: expected type `rustc::hir::def_id::DefId`
              found type `syntax::ast::NodeId`

error[E0308]: mismatched types
   --> clippy_lints/src/loops.rs:805:67
    |
805 |             let manual_copies = get_indexed_assignments(cx, body, def_id);
    |                                                                   ^^^^^^ expected struct `rustc::hir::def_id::DefId`, found struct `syntax::ast::NodeId`
    |
    = note: expected type `rustc::hir::def_id::DefId`
               found type `syntax::ast::NodeId`

error[E0308]: mismatched types
   --> clippy_lints/src/loops.rs:858:22
    |
858 |                 var: def_id,
    |                      ^^^^^^ expected struct `rustc::hir::def_id::DefId`, found struct `syntax::ast::NodeId`
    |
    = note: expected type `rustc::hir::def_id::DefId`
               found type `syntax::ast::NodeId`

error[E0308]: mismatched types
    --> clippy_lints/src/loops.rs:1678:35
     |
1678 |                 .as_local_node_id(def_id)
     |                                   ^^^^^^ expected struct `rustc::hir::def_id::DefId`, found struct `syntax::ast::NodeId`
     |
     = note: expected type `rustc::hir::def_id::DefId`
                found type `syntax::ast::NodeId`

error[E0308]: mismatched types
   --> clippy_lints/src/misc.rs:580:35
    |
580 |                 .as_local_node_id(def_id)
    |                                   ^^^^^^ expected struct `rustc::hir::def_id::DefId`, found struct `syntax::ast::NodeId`
    |
    = note: expected type `rustc::hir::def_id::DefId`
               found type `syntax::ast::NodeId`

error[E0308]: mismatched types
   --> clippy_lints/src/needless_pass_by_value.rs:133:38
    |
133 |                 !moved_vars.contains(&defid),
    |                                      ^^^^^^ expected struct `rustc::hir::def_id::DefId`, found struct `syntax::ast::NodeId`
    |
    = note: expected type `&rustc::hir::def_id::DefId`
               found type `&syntax::ast::NodeId`

error[E0308]: mismatched types
   --> clippy_lints/src/needless_pass_by_value.rs:142:59
    |
142 |                     let deref_span = spans_need_deref.get(&defid);
    |                                                           ^^^^^^ expected struct `rustc::hir::def_id::DefId`, found struct `syntax::ast::NodeId`
    |
    = note: expected type `&rustc::hir::def_id::DefId`
               found type `&syntax::ast::NodeId`

error: aborting due to 15 previous errors

error: Could not compile `clippy_lints`.

To learn more, run the command again with --verbose.
```
</details>

**Comment by oli-obk** [comment]
This is a fun one...

* definetely https://github.com/rust-lang/rust/pull/44316
* maybe https://github.com/rust-lang/rust/pull/44435
* possibly at runtime: https://github.com/rust-lang/rust/pull/44442

**Comment by oli-obk** [comment]
published as 0.0.159
